src/lib.rs

```diff
@@ -0,0 +1,408 @@
+#![forbid(unsafe_code)]
+
+mod baseline;
+mod check_release;
+mod config;
+mod dump;
+mod manifest;
+mod query;
+mod templating;
+mod util;
+
+use cargo_metadata::PackageId;
+pub use config::*;
+pub use query::*;
+
+use check_release::run_check_release;
+use trustfall_rustdoc::{load_rustdoc, VersionedCrate};
+
+use dump::RustDocCommand;
+use itertools::Itertools;
+use semver::Version;
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+
+/// Test a release for semver violations.
+pub struct Check {
+    /// Which packages to analyze.
+    scope: Scope,
+    current: Rustdoc,
+    baseline: Rustdoc,
+    log_level: Option<log::Level>,
+}
+
+pub struct Rustdoc {
+    source: RustdocSource,
+}
+
+impl Rustdoc {
+    /// Use an existing rustdoc file.
+    pub fn from_path(rustdoc_path: impl Into<PathBuf>) -> Self {
+        Self {
+            source: RustdocSource::Rustdoc(rustdoc_path.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the project root directory, i.e. the directory containing the crate source.
+    /// It can be a workspace or a single package.
+    /// Same as `from_git_revision`, but with the current git revision.
+    pub fn from_root(project_root: impl Into<PathBuf>) -> Self {
+        Self {
+            source: RustdocSource::Root(project_root.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the project at a given git revision.
+    pub fn from_git_revision(
+        project_root: impl Into<PathBuf>,
+        revision: impl Into<String>,
+    ) -> Self {
+        Self {
+            source: RustdocSource::Revision(project_root.into(), revision.into()),
+        }
+    }
+
+    /// Generate the rustdoc file from the largest-numbered non-yanked non-prerelease version
+    /// published to the cargo registry. If no such version, uses
+    /// the largest-numbered version including yanked and prerelease versions.
+    pub fn from_latest_version() -> Self {
+        Self {
+            source: RustdocSource::Version(None),
+        }
+    }
+
+    pub fn from_version(version: impl Into<String>) -> Self {
+        Self {
+            source: RustdocSource::Version(Some(version.into())),
+        }
+    }
+}
+
+enum RustdocSource {
```

Comment:
```suggestion
#[derive(Debug)]
enum RustdocSource {
```
for all types unless there's a reason not to?

Reply:
done :)